File: src/main/java/net/arvandor/numinoustreasury/measurement/Weight.java
package net.arvandor.numinoustreasury.measurement;

import static java.lang.Math.round;

import org.bukkit.configuration.serialization.ConfigurationSerializable;
import org.bukkit.configuration.serialization.SerializableAs;

import java.util.Map;

@SerializableAs("Weight")
public final class Weight implements Comparable<Weight>, ConfigurationSerializable {

    private final double value;
    private final WeightUnit unit;

    public Weight(double value, WeightUnit unit) {
        this.value = value;
        this.unit = unit;
    }

    public double getValue() {
        return value;
    }

    public WeightUnit getUnit() {
        return unit;
    }

    public Weight to(WeightUnit unit) {
        return new Weight((getValue() / getUnit().getScaleFactor()) * unit.getScaleFactor(), unit);
    }

    public Weight multiply(int amount) {
        return new Weight(getValue() * amount, getUnit());
    }

    @Override
    public int compareTo(Weight weight) {
        return (int) round((getValue() / getUnit().getScaleFactor()) - (weight.getValue() / weight.getUnit().getScaleFactor()));
    }

    @Override
    public String toString() {
        if (getValue() == (long) getValue()) {
            return String.format("%.0f%s", getValue(), getUnit().getName());
        } else {
            return String.format("%s%s", getValue(), getUnit().getName());
        }
    }

    @Override
    public Map<String, Object> serialize() {
        return Map.of(
                "value", getValue(),
                "unit", getUnit().getName()
        );
    }

    public static Weight deserialize(Map<String, Object> serialized) {
        return new Weight(
                ((Number) serialized.get("value")).doubleValue(),
                WeightUnit.getByName((String) serialized.get("unit"))
        );
    }
}

File: src/main/kotlin/net/arvandor/numinoustreasury/measurement/Weight.kt
package net.arvandor.numinoustreasury.measurement

import org.bukkit.configuration.serialization.ConfigurationSerializable
import org.bukkit.configuration.serialization.SerializableAs

@SerializableAs("Weight")
class Weight(val value: Double, val unit: WeightUnit) : Comparable<Weight>, ConfigurationSerializable {
    fun to(unit: WeightUnit): Weight {
        return Weight((value / this.unit.scaleFactor) * unit.scaleFactor, unit)
    }

    fun multiply(amount: Int): Weight {
        return Weight(value * amount, unit)
    }

    override fun compareTo(other: Weight): Int {
        return Math.round((value / unit.scaleFactor) - (other.value / other.unit.scaleFactor)).toInt()
    }

    override fun toString(): String {
        return if (value == value.toLong().toDouble()) {
            String.format("%.0f%s", value, unit.name)
        } else {
            String.format("%s%s", value, unit.name)
        }
    }

    override fun serialize(): Map<String, Any> {
        return mapOf(
            "value" to value,
            "unit" to unit.name,
        )
    }

    companion object {
        @JvmStatic
        fun deserialize(serialized: Map<String?, Any?>): Weight {
            return Weight(
                (serialized["value"] as Number).toDouble(),
                WeightUnit.getByName(serialized["unit"] as String)!!,
            )
        }
    }
}
